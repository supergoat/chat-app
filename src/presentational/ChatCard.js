// @flow
import React, { Component } from 'react';
import {
  Text,
  View,
  Image
} from 'react-native';
import ProfilePhoto from './ProfilePhoto';
import TimeStamp from './TimeStamp';
import NotificationBadge from './NotificationBadge';
import CAColors from '../common/CAColors';
import CAStyleSheet from '../common/CAStyleSheet';
import profilePhotoPlaceholder from '../assets/img/profile_photo_placeholder.png';

/* Constants
============================================================================= */
const CHAT_CARD_HEIGHT = 80;

/** ============================================================================
<ChatCard
  [profilePhoto]={string}
  name={string}
  latestMessage={{
    text: string,
    createdAt: Date
  }}
  unRead={number}
/>
--------------------------------------------------------------------------------
Displays information about a chat
@param { string } [profilePhoto] - The photo of the person the user is talking
@param { string } name - The name of the person the user is talking to
@param { object } latestMessage - The latest message of the chat
@param { string } unRead - The number of unRead messages
============================================================================= */
type Props = {
  profilePhoto?: string,
  name: string,
  latestMessage: {
    text: string,
    createdAt: Date
  },
  unRead: number
};
class ChatCard extends Component<Props> {
  static __cards__: Function;

  static defaultProps = {
    unRead: 0
  };

  render() {
    const { profilePhoto, name, latestMessage, unRead } = this.props;
    return (
      <View style={styles.chatCard}>
        <ProfilePhoto
          profilePhoto={profilePhoto}
          name={name}
          style={styles.profilePhoto}
        />

        <View style={styles.chatInfo}>
          <View style={styles.heading}>
            <Text style={styles.name}>{name}</Text>

            <TimeStamp
              date={latestMessage.createdAt}
              style={unRead > 0 ? styles.timeStamp : {}}
            />
          </View>

          <View style={styles.body}>
            <Text
              style={styles.latestMessage}
              ellipsizeMode="tail"
              numberOfLines={1}
            >
              {latestMessage.text}
            </Text>

            <NotificationBadge
              noOfNotifications={unRead}
              style={styles.notificationBadge}
            />
          </View>
        </View>
      </View>
    );
  }
}

/* StyleSheet
============================================================================= */
const styles = CAStyleSheet.create({
  chatCard: {
    height: CHAT_CARD_HEIGHT,
    backgroundColor: CAColors.white,
    alignItems: 'center',
    flexDirection: 'row'
  },
  profilePhoto: {
    marginLeft: 15
  },
  chatInfo: {
    height: CHAT_CARD_HEIGHT,
    flex: 1,
    margin: 10,
    borderBottomWidth: 1,
    borderColor: CAColors.gallery,
    justifyContent: 'center'
  },
  heading: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 5
  },
  body: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center'
  },
  name: {
    fontWeight: 'bold'
  },
  timeStamp: {
    color: CAColors.forestGreen
  },
  latestMessage: {
    color: CAColors.oxfordBlue,
    flex: 0.95
  }
});

/* Playground Cards - the different states of ChatCard
============================================================================= */
ChatCard.__cards__ = define => {
  let anHourAgo = new Date();
  anHourAgo.setHours(anHourAgo.getHours() - 1);

  let aDayAgo = new Date();
  aDayAgo.setDate(aDayAgo.getDate() - 1);

  let twoDaysAgo = new Date();
  twoDaysAgo.setDate(twoDaysAgo.getDate() - 2);

  define('Simple', () => {
    return (
      <ChatCard
        profilePhoto="https://placekitten.com/60/60"
        name="Alex Stone"
        latestMessage={{
          text: 'Hey whats up?',
          createdAt: aDayAgo
        }}
        unRead={10}
      />
    );
  });

  define('With notification over 99', () => {
    return (
      <ChatCard
        name="Alex Stone"
        latestMessage={{
          text: 'Hey man how is it going? Long time no see.',
          createdAt: anHourAgo
        }}
        unRead={100}
      />
    );
  });

  define('With truncated latest message text', () => {
    return (
      <ChatCard
        name="Alex Stone"
        latestMessage={{
          text: 'Hey man how is it going? Long time no see.',
          createdAt: aDayAgo
        }}
      />
    );
  });

  define('With the date as a timestamp and a notification', () => {
    return (
      <ChatCard
        name="Alex Stone"
        latestMessage={{
          text: 'Hey whats up?',
          createdAt: twoDaysAgo
        }}
        unRead={10}
      />
    );
  });
};

/* Export
============================================================================= */
export default ChatCard;
