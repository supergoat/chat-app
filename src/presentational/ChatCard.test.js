// @flow
import React from 'react';
import ChatCard from './ChatCard';
import renderer from 'react-test-renderer';

describe('ChatCard', () => {

  // Custom constant date, so our snapshots don't fail when the date changes
  const customDate = new Date(2018, 1, 3, 15, 30);
  const RealDate = Date;

  beforeEach(() => {
    // Mock Date to always return our custom date
    global.Date = class extends RealDate {
      constructor() {
        return new RealDate(customDate);
      }
    };
  });

  afterEach(() => {
    // Set the Date back to the real date
    global.Date = RealDate;
  });

  it('displays the profilePhoto when it is provided', () => {
    const tree = renderer.create(
      <ChatCard
        profilePhoto='https://placekitten.com/60/60'
        name="Alex Stone"
        latestMessage={{
          text: "Hey whats up?",
          createdAt: new Date()
        }}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('it displays a profilePhoto placeholder when it is NOT provided', () => {
    const tree = renderer.create(
      <ChatCard
        name="James Gordon"
        latestMessage={{
          text: "Hey how is it going?",
          createdAt: new Date()
        }}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('truncates the latest message text when its longer than one line', () => {
    const tree = renderer.create(
      <ChatCard
        name="James Gordon"
        latestMessage={{
          text: "Hey man how is it going? Long time no see.",
          createdAt: new Date()
        }}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays a notification when there are undread messages', () => {
    const tree = renderer.create(
      <ChatCard
        name="James Gordon"
        latestMessage={{
          text: "Hey man how is it going? Long time no see.",
          createdAt: new Date()
        }}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
