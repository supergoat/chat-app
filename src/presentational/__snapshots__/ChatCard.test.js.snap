// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatCard displays a notification when there are undread messages 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flexDirection": "row",
      "height": 80,
    }
  }
>
  <Image
    alt="James Gordon"
    source={
      Object {
        "testUri": "../../../src/assets/img/profile_photo_placeholder.png",
      }
    }
    style={
      Array [
        Object {
          "backgroundColor": "#EEEEEE",
          "borderRadius": 25,
          "height": 50,
          "width": 50,
        },
        Object {
          "marginLeft": 15,
        },
      ]
    }
  />
  <View
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#EEEEEE",
        "flex": 1,
        "height": 80,
        "justifyContent": "center",
        "margin": 10,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 5,
        }
      }
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        James Gordon
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#333C4E",
              "fontSize": 12,
            },
            Object {},
          ]
        }
      >
        15:30
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "color": "#333C4E",
            "flex": 0.95,
          }
        }
      >
        Hey man how is it going? Long time no see.
      </Text>
    </View>
  </View>
</View>
`;

exports[`ChatCard displays the profilePhoto when it is provided 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flexDirection": "row",
      "height": 80,
    }
  }
>
  <Image
    alt="Alex Stone"
    source={
      Object {
        "uri": "https://placekitten.com/60/60",
      }
    }
    style={
      Array [
        Object {
          "backgroundColor": "#EEEEEE",
          "borderRadius": 25,
          "height": 50,
          "width": 50,
        },
        Object {
          "marginLeft": 15,
        },
      ]
    }
  />
  <View
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#EEEEEE",
        "flex": 1,
        "height": 80,
        "justifyContent": "center",
        "margin": 10,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 5,
        }
      }
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Alex Stone
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#333C4E",
              "fontSize": 12,
            },
            Object {},
          ]
        }
      >
        15:30
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "color": "#333C4E",
            "flex": 0.95,
          }
        }
      >
        Hey whats up?
      </Text>
    </View>
  </View>
</View>
`;

exports[`ChatCard it displays a profilePhoto placeholder when it is NOT provided 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flexDirection": "row",
      "height": 80,
    }
  }
>
  <Image
    alt="James Gordon"
    source={
      Object {
        "testUri": "../../../src/assets/img/profile_photo_placeholder.png",
      }
    }
    style={
      Array [
        Object {
          "backgroundColor": "#EEEEEE",
          "borderRadius": 25,
          "height": 50,
          "width": 50,
        },
        Object {
          "marginLeft": 15,
        },
      ]
    }
  />
  <View
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#EEEEEE",
        "flex": 1,
        "height": 80,
        "justifyContent": "center",
        "margin": 10,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 5,
        }
      }
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        James Gordon
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#333C4E",
              "fontSize": 12,
            },
            Object {},
          ]
        }
      >
        15:30
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "color": "#333C4E",
            "flex": 0.95,
          }
        }
      >
        Hey how is it going?
      </Text>
    </View>
  </View>
</View>
`;

exports[`ChatCard truncates the latest message text when its longer than one line 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flexDirection": "row",
      "height": 80,
    }
  }
>
  <Image
    alt="James Gordon"
    source={
      Object {
        "testUri": "../../../src/assets/img/profile_photo_placeholder.png",
      }
    }
    style={
      Array [
        Object {
          "backgroundColor": "#EEEEEE",
          "borderRadius": 25,
          "height": 50,
          "width": 50,
        },
        Object {
          "marginLeft": 15,
        },
      ]
    }
  />
  <View
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#EEEEEE",
        "flex": 1,
        "height": 80,
        "justifyContent": "center",
        "margin": 10,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 5,
        }
      }
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        James Gordon
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#333C4E",
              "fontSize": 12,
            },
            Object {},
          ]
        }
      >
        15:30
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "color": "#333C4E",
            "flex": 0.95,
          }
        }
      >
        Hey man how is it going? Long time no see.
      </Text>
    </View>
  </View>
</View>
`;
